File: exercism/python/luhn/luhn.py

class Luhn(object):

    MAPPING = dict(zip('0123456789', '0246813579'))

    def __init__(self, given):
        self.digits = list(str(given))

    def addends(self):
        kept = self.digits[-1::-2]
        mapped = [self.MAPPING[d] for d in self.digits[-2::-2]]
        return [int(d) for d in kept + mapped]

    def checksum(self):
        return sum(self.addends())

    def is_valid(self):
        return self.checksum() % 10 == 0

    @staticmethod
    def create(given):
        candidate = given * 10
        return candidate + (-Luhn(candidate).checksum() % 10)

File: exercism/haskell/luhn/Luhn.hs
module Luhn (checkDigit, addends, checksum, isValid, create) where

import Data.List (unfoldr)
import qualified Data.Map as M

checkDigit :: Integer -> Integer
checkDigit = (`mod` 10)

addends :: Integer -> [Integer]
addends n = reverse $ unfoldr f (n, 0)
  where f (0, _) = Nothing
        f (m, par) = Just (conv par (m `rem` 10), (m `div` 10, (par + 1) `mod` 2))
        conv par = (M.!) (mappings !! par)
        sing = M.fromList $ zip [0..9] [0..9]
        doub = M.fromList $ zip [0..9] [0, 2, 4, 6, 8, 1, 3, 5, 7, 9]
        mappings = [sing, doub]

checksum :: Integer -> Integer
checksum = (`mod` 10) . sum . addends

isValid :: Integer -> Bool
isValid n = checksum n `mod` 10 == 0

create :: Integer -> Integer
create n = n * 10 + (- checksum (n * 10)) `mod` 10
